### DISCUSSION TITLE
ml-matrix Matrix constructor fails

### DISCUSSION THREAD

**Discussion by kletts** [observablehq/framework#1221] [Q&A] [answered]
I'm trying to use the matrix operations library from [ml-matrix](https://www.npmjs.com/package/ml-matrix) in an observable framework notebook for which the new Matrix constructor is returning the following error: 

```
TypeError: Cannot read properties of null (reading 'isAnyArray')
```

The following code reproduces the error. The first two examples work, showing the import of the library was successful. The third example fails. Testing in node.js and in an online observable notebook the third example is working. 

```js
import { Matrix } from "npm:ml-matrix";
const mm1 = Matrix.ones(5, 5);
display(mm1);
const mm2 = new Matrix(2,2); 
display(mm2);
const mm3 = new Matrix([
  [1, 1],
  [2, 2],
]);
display(mm3);
```
ml-matrix does have a dependency on [is-any-array](https://www.npmjs.com/package/is-any-array) 

Any help would be appreciated. 

Christian

**Comment by mbostock** [comment, marked as answer]
You can fix this by changing how you import ml-matrix like so:

```js
import {Matrix} from "npm:ml-matrix/src/index.js/+esm";
```

The problem is that the default entry point for ml-matrix is [matrix.mjs](https://cdn.jsdelivr.net/npm/ml-matrix/matrix.mjs), which imports [matrix.js](https://cdn.jsdelivr.net/npm/ml-matrix/matrix.js), which is CommonJS [generated by rollup](https://github.com/mljs/matrix/blob/main/rollup.config.mjs). Unfortunately converting ES modules to CommonJS and back again loses information. If you look at the generated matrix.js, it imports is-any-array like this:

```js
var isAnyArray = require('is-any-array');
```

and then references `isAnyArray.isAnyArray`. So when jsDelivr tries to convert this CommonJS back to an ES module, you get:

```js
import t from"/npm/is-any-array@2.0.1/+esm";
```

and `t.isAnyArray`. But here `t` is the default export, not a namespace, so you get the error. The correct import of is-any-array is seen  if you tell jsDelivr to produce the ES module bundle from the ES module source, ignoring ml-matrix’s CommonJS bundle.

```js
import{isAnyArray as t}from"/npm/is-any-array@2.0.1/+esm";
```

Anyway, I’m sure this is all just annoying noise to you as the user of ml-matrix. But the real answer is that ml-matrix needs to fix its package.json so that it’s distributing valid ES modules, rather than an ES module that imports CommonJS.

**Reply by kletts** [reply]
Oh excellent thank you

**Reply by Fil** [reply]
to celebrate this I ported the `pt` utility to Framework:
https://observablehq.observablehq.cloud/pangea/varia/pt
